qt-models/diveplannermodel.cpp

```diff
@@ -537,14 +588,16 @@ int DivePlannerPointsModel::gfLow() const
 	return diveplan.gflow;
 }
 
-void DivePlannerPointsModel::setRebreatherMode(int mode)
+void DivePlannerPointsModel::setDiveMode(int mode)
```

Comment:
This function looks like a layering violation. The model shouldn't modify the dive. I will have to untangle this.

Reply:
> This function looks like a layering violation. The model shouldn't modify the dive. I will have to untangle this.

I think the whole layering is somewhat weird around the DivePlannerPointsModel, as this 'owns' a Dive (that isn't part of the dive list either). Besides the layering violation this is also creating a lot of headaches coming from the 'dual use' of the Dive structure.
But if we want to refactor this, can I propose that we do this outside of this pull request that is intended to bring a quite substantial functional change?

Reply:
> I think the whole layering is somewhat weird around the DivePlannerPointsModel, as this 'owns' a Dive (that isn't part of the dive list either). 

It doesn't own the dive - it only holds a reference to it. The owners are `PlannerWidgets` (for planned dives) and `ProfileWidget` (for edited dives) respectively. The model shouldn't modify the dive (yes, I am painfully aware of `createSimpleDive()`).

I would suggest something like this, i.e. let the owner manipulate the dive and then inform the model of the change (totally untested):
```
diff --git a/desktop-widgets/diveplanner.cpp b/desktop-widgets/diveplanner.cpp
index 64ce4758a..774b5cecc 100644
--- a/desktop-widgets/diveplanner.cpp
+++ b/desktop-widgets/diveplanner.cpp
@@ -555,7 +555,6 @@ PlannerWidgets::PlannerWidgets() :
 	plannerSettingsWidget(this)
 {
 	plannerWidget.setDiveMode(OC);
-	//setDiveMode(OC);
 
 	connect(plannerDetails.printPlan(), &QPushButton::pressed, this, &PlannerWidgets::printDecoPlan);
 	connect(DivePlannerPointsModel::instance(), &DivePlannerPointsModel::calculatedPlanNotes,
@@ -680,7 +679,11 @@ void PlannerWidgets::printDecoPlan()
 
 void PlannerWidgets::setDiveMode(int mode)
 {
-	DivePlannerPointsModel::instance()->setDiveMode(mode);
+	struct divecomputer *dc = planned_dive->get_dc(dcNr);
+	if (dc)
+		dc->divemode = (divemode_t) mode;
+
+	DivePlannerPointsModel::instance()->cylindersChanged();
 	plannerWidget.setColumnVisibility(mode);
 	plannerSettingsWidget.setBailoutVisibility(mode);
 }
diff --git a/qt-models/diveplannermodel.cpp b/qt-models/diveplannermodel.cpp
index f203d772f..a84c0e587 100644
--- a/qt-models/diveplannermodel.cpp
+++ b/qt-models/diveplannermodel.cpp
@@ -588,12 +588,10 @@ int DivePlannerPointsModel::gfLow() const
 	return diveplan.gflow;
 }
 
-void DivePlannerPointsModel::setDiveMode(int mode)
+void DivePlannerPointsModel::cylindersChanged()
 {
-	struct divecomputer *dc = d->get_dc(dcNr);
-	if (dc)
-		dc->divemode = (divemode_t) mode;
-
+	if (!d)
+		return;
 	cylinders.updateDive(d, dcNr);
 
 	emitDataChanged();
diff --git a/qt-models/diveplannermodel.h b/qt-models/diveplannermodel.h
index a3bc25bd5..9acdc7660 100644
--- a/qt-models/diveplannermodel.h
+++ b/qt-models/diveplannermodel.h
@@ -66,7 +66,7 @@ public:
 
 	void loadFromDive(dive *d, int dcNr);
 	void addStop(int millimeters, int seconds);
-	void setDiveMode(int mode);
+	void cylindersChanged();
 public
 slots:
 	void addDefaultStop();
```

I think I tried to remove the `DivePlannerPointsModel` global a number of times, but gave up every time. Let's wait for a rainy afternoon, when I find more time...